src/senaite/core/adapters/localroles.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of SENAITE.CORE.
+#
+# SENAITE.CORE is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright 2018-2022 by it's authors.
+# Some rights reserved, see README and LICENSE.
+
+import six
+from bika.lims import api
+from bika.lims import logger
+from borg.localrole.default_adapter import DefaultLocalRoleAdapter
+from senaite.core.interfaces import IDynamicLocalRoles
+from zope.component import getAdapters
+
+
+class DynamicLocalRoleAdapter(DefaultLocalRoleAdapter):
+    """Gives additional Member local roles based on current user and context
+    This enables giving additional permissions on items out of the user's
+    current traverse path
+    """
+
+    _roles_in_context = {}
+
+    def getRolesInContext(self, context, principal_id):
+        """Returns the dynamically calculated 'local' roles for the given
+        principal and context
+        @param context: context to calculate roles for the given principal
+        @param principal_id: User login id
+        @return List of dynamically calculated local-roles for user and context
+        """
+        if not api.is_object(context):
+            # We only apply dynamic local roles to valid objects
+            return []
+
+        # This function is called a lot within same request, do some cache
+        context_uid = api.get_uid(context)
+        roles = self._roles_in_context.get(context_uid, {})
+        if principal_id in roles:
+            return roles.get(principal_id)
+
+        # Look for adapters
+        roles = []
+        path = api.get_path(context)
+        adapters = getAdapters((context,), IDynamicLocalRoles)
```

Comment:
What do you think about making it a multi adapter for `context, request`?
Then we could probably more easy override default adapters later.

Reply:
True, but then we need to add a `priority` attribute to prioritize adapters. We won't be able to override the roles inferred by other adapters with our new one otherwise.

Reply:
Aren't these role adapters all additive?
Therefore, I don't think that we need a priority here.
Actually I don't understand what you mean by `We won't be able to override the roles inferred by other adapters with our new one otherwise.`

Reply:
Yes, these adapters are all additive. 

> Therefore, I don't think that we need a priority here.
> Actually I don't understand what you mean by `We won't be able to override the roles inferred by other adapters with our new one otherwise.`

Yes, you are right, apologies. If we wanted to "remove" roles added dynamically by adapters, we don't have option other than overriding this `DynamicLocalRoleAdapter` itself. Because the adapters are additive, precisely.